### PR TITLE
Periodic Substation Maintainance

### PR DESCRIPTION
**Pull request by joshepvodka**
Inspired by #18278 I did this WIP code to add the need for periodic substation maintenance to the game. It's my 1st commit and the code is very barebones so bare with me please. I plan to keep working on it if feedback is positive.

Changelog:

**Substations now will wear down over time:**
Substations now have an Integrity value that decreases over time according to how much power the substation is supplying. When Integrity reaches 0% the substation will Malfunction, disabling it until someone comes along to fix it. Substations cannot malfunction due to over time wear in between 5 minutes of each other.

**Added substation "maintenance":**
To fix a malfunctioning substation someone needs to disassemble it into it's machine frame stage and screwdriver it back together. This was done as a proof of concept, for now I don't know how to implement a less bare bones way to fix it, but definitely plan to make something better.

**Substations can now be EMAGed:**
EMAGing a substation will make it malfunction as if from normal use, same way to fix it applies.

Some insight about these changes:

I tried doing this with as little code as possible, only added SubstationComponent and SubstationSystem for now as to not make the changes bloated with poorly written code.

When a substation is made malfunction by the SubstationSystem the loss of integrity is halted for 5 minutes, as not to make have a wave of blackouts that require maints diving happen.

Due to wear being proportional to use, departments that need plenty of electricity will have their substations malfunction first, from what I got by testing it, Science and Med will usually go out first. (But again, at the very most 5 minutes from each other)

The Integrity of substations is visible as a percentage through the examine window. When it malfunctions the examine window also shows that. Also when a station malfunctions I remove the ExaminableBatteryComponent as not to show unneeded information, I don't know if that's an alright or bad practice, for now it works fine but certainly can be buggy in future versions.

Substations can now be EMAGed because it was easy code to implement and I thought it made sense. Should make sabotage less apparent than cutting the HV cables to the substation. I fully understand if this feature is not wanted or needed.

Substations are fixed through reassembly because, as of now, I simply don't know how to implement a better fleshed out way, but again, I plan of working on it.

Wallmount Substations remain unchanged as to not let the emergency or arrivals shuttle have a blackout.

Ideas:

Substation maintenance should stay simple and easy, but probably should involve opening the wire panel and changing a component, like a capacitor, breaker or fuse. (something easily to get through a lathe or an electronics vendor machine.

Making wear be somewhat random as every round different substations fail at different times.

image
image

I'll gladly take any feedback, positive or negative, about the code or the ideas I'm putting out here.

### CODE REVIEW

**Comment by UbaserB** [comment]
i like this idea

**Review by keronshb — CHANGES_REQUESTED** [review]
Hey thanks for your first PR! I haven't been able to sit down for a full proper review but I did notice something right away.

> Substations can now be EMAGed:
EMAGing a substation will make it malfunction as if from normal use, same way to fix it applies.

It doesn't say this anywhere now so there's absolutely no way you would've known but can you please take out the emag interaction? There's an over-reliance on emag interactions and we're trying to cut down on it. I'm going to make a freeze for it right now.

**Comment by metalgearsloth** [comment]
Did you see notafet / partmedia's implementation?

**Comment by joshepvodka** [comment]
> Did you see notafet / partmedia's implementation?

Yes I have! I thought it was a pretty cool idea, just outside my current skill bracket right now. Though also I worry about adding too much on top of it when I (personally) think substation maintenance should be just something to drive away the boredom of a slow engineering shift. I'll be glad to take it that way though if people think it would be more fun

**Comment by joshepvodka** [comment]
> Hey thanks for your first PR! I haven't been able to sit down for a full proper review but I did notice something right away.
> 
> > Substations can now be EMAGed:
> > EMAGing a substation will make it malfunction as if from normal use, same way to fix it applies.
> 
> It doesn't say this anywhere now so there's absolutely no way you would've known but can you please take out the emag interaction? There's an over-reliance on emag interactions and we're trying to cut down on it. I'm going to make a freeze for it right now.

Done, also commited a missing file along with it

**Review by keronshb — APPROVED** [review]

**Comment by daerSeebaer** [comment]
I like the idea - gives engineering something to do, they can do preventive maintenance to control the situation or have to run if they forgot (or had too much to do?), scaling with power makes this kind of predictable enough to not make it frustrating.

Did I calculate it correctly that each Substation has 30MJ of lifetime? That would be about 20 mins for Science and multiple hours for low-use Substations, which feels like a good spot.

One thing that I miss here tough is the ability to turn this off for individual substations. There are Map files that use normal substations where decay could be devastating (i.e. the Cargo shuttle) .
In the same thought: You introduced a new Component variable "Enabled" that I don't quite understand. Does this mean the Substation is enabled? Or that the wear feature is enabled?
If it means the Substation is enabled it should probably be more integrated with the Substation power components.
But my suggestion would be to just rename this "decayEnabled" and make it VV read-writeable, so mappers can make individual Substations immune to decay and we don't have a redundant variable.

As another step I would start exactly that discussion and ask everyone who's reading this which Substations should have decay enabled and which not. My suggestion: Disable it on everything but the regular stations just to be sure.

And if I read the code correctly, right now the 5-minute delay is not yet implemented? One way to do it is to make an attribute like "endOfDecayTimeOut". When a Substation breaks, set it to now+5 minutes. And in your Update() function, first thing check if _substationWearEnabled is false AND the current game Time is greater than endOfDecayTimeOut and re-enable it if yes

**Comment by BolloTea** [comment]
Maybe something similar can be done with security cameras?

**Comment by chromiumboy** [comment]
Overall I like the idea of this mechanic and would like to see how it goes

That said

> One thing that I miss here tough is the ability to turn this off for individual substations. There are Map files that use normal substations where decay could be devastating (i.e. the Cargo shuttle)

This is a very important point. CentComm, thr arrivals station, shuttle substations, anything off station really, needs to be protected from this mechanic

Also there should be a distinct visual / sound effect to indicate that a substation is malfunctioning (if there isn't one already)

**Comment by metalgearsloth** [comment]
> This is a very important point. CentComm, thr arrivals station, shuttle substations, anything off station really, needs to be protected from this mechanic

> Also there should be a distinct visual / sound effect to indicate that a substation is malfunctioning (if there isn't one already)

Maybe some kind of T2 / T3 research for an APC that doesn't decay? That way it's a mechanical thing ingame rather than mapper magic.

**Comment by chromiumboy** [comment]
> Maybe some kind of T2 / T3 research for an APC that doesn't decay? That way it's a mechanical thing ingame rather than mapper magic.

I'd be down for that

Could instead be a benefit bestowed by upgrading the substation with a full set of bluespace capacitors? But that might too hard to convey in game

An upgrade kit for substations that you can attach? Then you could get engineers to upgrade primary substations which get a lot of wear so they don't need maintenance. Would be good for long rounds

**Comment by Ilya246** [comment]
i'm fairly sure this will cause a considerable amount of containment failure singulooses

**Comment by UbaserB** [comment]
> i'm fairly sure this will cause a considerable amount of containment failure singulooses

LOL TRUE (not a particularly bad thing though)

**Comment by chromiumboy** [comment]
> i'm fairly sure this will cause a considerable amount of containment failure singulooses

Another good reason to include a visual and sound fx, as the substation that powers to singulo containment emitters should be in the heart of engineering. Any connected APCs will also lose power before the emitters fail most likely

**Comment by KompaniLinge** [comment]
Will this be viewable via the power monitoring computer?

**Comment by Ilya246** [comment]
> > i'm fairly sure this will cause a considerable amount of containment failure singulooses
> 
> Another good reason to include a visual and sound fx, as the substation that powers to singulo containment emitters should be in the heart of engineering. Any connected APCs will also lose power before the emitters fail most likely

emits are MV and also seem to have a property where they take a while to react to being enabled or power coming back on unless manually rebooted

**Comment by chromiumboy** [comment]
> Will malfunctions be viewable via the power monitoring computer? As an eng main i can assure you that would make this a lot more fun or interesting instead of just staring at wires until you can track down wether a substation is malfunctioning, or you got a wirecutter or not.

If/when this is completed I'll look into it

**Comment by joshepvodka** [comment]
> I like the idea - gives engineering something to do, they can do preventive maintenance to control the situation or have to run if they forgot (or had too much to do?), scaling with power makes this kind of predictable enough to not make it frustrating.
> 
> Did I calculate it correctly that each Substation has 30MJ of lifetime? That would be about 20 mins for Science and multiple hours for low-use Substations, which feels like a good spot.
> 
> One thing that I miss here tough is the ability to turn this off for individual substations. There are Map files that use normal substations where decay could be devastating (i.e. the Cargo shuttle) . In the same thought: You introduced a new Component variable "Enabled" that I don't quite understand. Does this mean the Substation is enabled? Or that the wear feature is enabled? If it means the Substation is enabled it should probably be more integrated with the Substation power components. But my suggestion would be to just rename this "decayEnabled" and make it VV read-writeable, so mappers can make individual Substations immune to decay and we don't have a redundant variable.
> 
> As another step I would start exactly that discussion and ask everyone who's reading this which Substations should have decay enabled and which not. My suggestion: Disable it on everything but the regular stations just to be sure.
> 
> And if I read the code correctly, right now the 5-minute delay is not yet implemented? One way to do it is to make an attribute like "endOfDecayTimeOut". When a Substation breaks, set it to now+5 minutes. And in your Update() function, first thing check if _substationWearEnabled is false AND the current game Time is greater than endOfDecayTimeOut and re-enable it if yes

Thanks for your input! I just added the DecayEnabled variable to the SubstationComponent, i didnt realize before i could do that, that idea is really essencial.
Also, yeah, I messed up when coding and completely forgot to check if the timer was working, now it should be working correctly.
Also x2, because the DecayEnabled variable has been added and now specific substations can be set to never malfunction i've added the component to all substations, not just the SubstationBasic entity

**Comment by joshepvodka** [comment]
> Will malfunctions be viewable via the power monitoring computer? As an eng main i can assure you that would make this a lot more fun or interesting instead of just staring at wires until you can track down wether a substation is malfunctioning, or you got a wirecutter or not.

During my testing i've found out that malfunctioning substations are listed as consuming 0W on the power monitor. I'd be glad to add a "Malfunctioning!" warning or integrity value on the power monitor ui as soon as i find out and if this idea gets support.

**Comment by joshepvodka** [comment]
> > Maybe some kind of T2 / T3 research for an APC that doesn't decay? That way it's a mechanical thing ingame rather than mapper magic.
> 
> I'd be down for that
> 
> Could instead be a benefit bestowed by upgrading the substation with a full set of bluespace capacitors? But that might too hard to convey in game
> 
> An upgrade kit for substations that you can attach? Then you could get engineers to upgrade primary substations which get a lot of wear so they don't need maintenance. Would be good for long rounds

I support these ideas, i think it's really important that players are able to build everything a mapper can if they'd like. I think capacitors extending substation lifetime is a good place to start, though maybe cap it out at tier 3 capacitors as bluespace ones are very hard to come by.

**Comment by joshepvodka** [comment]
> > i'm fairly sure this will cause a considerable amount of containment failure singulooses
> 
> Another good reason to include a visual and sound fx, as the substation that powers to singulo containment emitters should be in the heart of engineering. Any connected APCs will also lose power before the emitters fail most likely

I'd very much like to implement visual and sfx for when substations malfunction, but for now I don't know what those would be. I was thinking maybe a emp sound for when the station fails? any suggestions?

**Comment by chromiumboy** [comment]
> > > i'm fairly sure this will cause a considerable amount of containment failure singulooses
> > 
> > 
> > Another good reason to include a visual and sound fx, as the substation that powers to singulo containment emitters should be in the heart of engineering. Any connected APCs will also lose power before the emitters fail most likely
> 
> I'd very much like to implement visual and sfx for when substations malfunction, but for now I don't know what those would be. I was thinking maybe a emp sound for when the station fails? any suggestions?

The screen could go black, like it's unanchored, and sparks could be thrown off of it with accompanying sparking sounds, maybe

**Comment by daerSeebaer** [comment]
> The screen could go black, like it's unanchored

If we want engineers to have to take a closer look, sure. If we want to accommodate them the screen could indicate a malfunction by going red, or flashing, or both

**Comment by joshepvodka** [comment]
Now when substations get down to 70% integrity it's screen and light will go yellow and the light will start blinking. When it goes down to 30% both will go red and the light will keep blinking. I hope I can finish this and get it into a proper PR after chromiumboy gets their power monitor overhaul merged, so i can make malfunctioning substations show up in the computer's map.

I still haven't gotten around to making a proper way to restore integrity to stations, and I'm out of ideas, maybe some sort of refueling? Space lube, possible "liquid conductor", some coolant glass or liquid or maybe just a single fuse item? Can i get some opinions?

**Comment by daerSeebaer** [comment]
> I still haven't gotten around to making a proper way to restore integrity to stations

So the boring way would be to do it like damaged windows - hold a welder to it. Easy to implement, not really engaging.

My suggestion would be to stay with the reconstruction theme, but spice it up by the Substation needing some parts replaced:
- All parts needed for a Substation are available on all stations, so this "just works" and you don't need to implement additional items (and make them available on all maps)
- It's basically refueling, but using an existing system
- Rebuilding is more engaging and even has a short learning curve (experienced engineers can do it faster)
- By needing to replace parts, there is still a material cost to it and a minimal amount of planning / inventory space required

My suggestion with some implementation thoughts (assuming Machine construction works as I think it does): On every 20% integrity step (80, 60, ...) delete a MV cable from the underlying ConstructibleMachine (there are 5 in it, so on machine failure the last one will go). When an engineer rebuilds the Substation, they have to replace the missing cables.
- MV cables are available in engineering and can be easily reproduced, so no risk of crippling a station through material requirements
- preventive maintenance is even more rewarding since it scales, giving engineers another reason to not let the substations run out
- should be somewhat easy to implement
- maybe adjust the warning screen threshold to indicate after the first lost cable (80%) so engis know they can't lossless deconstruct this one
- having MV cables in a substation get worn out and having to fix something _in_ the machine feels fitting in-universe
- _finally a use for all those MV cables_

**Comment by daerSeebaer** [comment]
What do you think about coupling the lifetime of the Substation to it's capacity? That way upgrading Substations with better parts actually makes engis life easier. And if we go with rebuilding as way to repair, you are fiddling with the parts anyway.

**Comment by joshepvodka** [comment]
Big update, added stuff:
New item (missing sprite): HV Fuse
Now substations require a fuse to work, they can be slotted in or changed by first opening the panel of the substation.
Fuses should be filled up a 50:50 mix of plasma and nitrogen, to act as a "protective gas" for the filament, the contents will slowly be degraded into NO2 as the substations works.
Now the substation integrity is tied to the amount of N2 and Plasma inside the fuses.
Fuses are craftable through the autolathe from roundstart, requiring 1 steel and 1 glass.

Fuses should give atmos some maintenance duty to do, be it just filling up fuses the other engineers get them or going around with a canister maintaining substations.
Also, no matter how much gas its crammed into the fuse they will always be consumed at the same rate, to change that engineers will need to instead upgrade substations.

Fuse lifetime upgrades:
Upgrading substations with better capacitors will now make fuses last longer. T2 capacitors will double their lifetime ad T3 will make it indefinite.

Other changes:
added the SubstationWallBasicEmpty prototype as the wallmounted counterpart for SubstationBasicEmpty. Now when constructing wall substations manualy it should be spawned in.
now wallmounted substations have WirespanelComponents, so you can open them up and change their fuses.
added AddMaxUpgrade function to UpgradeExamineEvent so players can see when the fuse lifetime is fully upgraded.

Footnotes:
(As a frequent atmos player) I feel like atmos has a very passive role in the station, so that's why i chose to make substation maintenance have gas tanks involved and not material replacements or oil changes.
I feel like this PR is nearly ready to be closed, I've added everything I wanted to add, but I'll gladly change stuff until reviewers are satisfied with it.
For the changed to actually be merged maps will need to be changed, disabling decay or upgrading specific substations, such as the ones in the terminal, shuttles and planets.

File: Content.Client/Power/Substation/SubstationVisualsSystem.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+namespace Content.Client.Power.Substation;
+
+public sealed class SubstationVisualsSystem : VisualizerSystem<SubstationVisualsComponent>
+{
+    protected override void OnAppearanceChange(EntityUid uid, SubstationVisualsComponent component, ref AppearanceChangeEvent args)
+    {
+        if (args.Sprite == null || !args.Sprite.LayerMapTryGet(component.LayerMap, out var layer))
+            return;
+
+		if(args.AppearanceData.TryGetValue(SubstationVisuals.Screen, out var stateObject)
```
Comment (0x6273) [inline Content.Client/Power/Substation/SubstationVisualsSystem.cs:15]:
```suggestion
        if (args.Sprite == null || !args.Sprite.LayerMapTryGet(component.LayerMap, out var layer))
            return;

		if (args.AppearanceData.TryGetValue(SubstationVisuals.Screen, out var stateObject)
```
Second if should have space before the parentheses, like the first if. Fix this in all the files, not just this one.

File: Content.Server/Power/Components/SubstationComponent.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    [DataField("decayEnabled")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public bool DecayEnabled = true;
+   
+    [DataField("state")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public SubstationIntegrityState State = SubstationIntegrityState.Healthy;
+
+    //9.231205828 is the amount of moles in a 5L container (the default fuse) at 1000Kpa 20C°
+    public float initialFuseMoles = 2.051379050f;
```
Comment (0x6273) [inline Content.Server/Power/Components/SubstationComponent.cs:21]:
```suggestion
    public float InitialFuseMoles = 2.051379050f;
```
Fields should be PascalCase.

File: Content.Client/Power/Substation/SubstationVisualsComponent.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace Content.Client.Power.Substation;
+
+[RegisterComponent]
+public sealed partial class SubstationVisualsComponent : Component
+{
+    [DataField("screenLayer")]
+    public string LayerMap { get; private set; } = string.Empty;
+
+    [DataField("integrityStates")]
```
Comment (0x6273) [inline Content.Client/Power/Substation/SubstationVisualsComponent.cs:11]:
```suggestion
    [DataField]
```
DataField names now auto generate from the field name. You only need to specify it if you need the DataField name to be different from the class field name (which it usually should not be).

File: Content.Server/Power/EntitySystems/SubstationSystem.cs [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+public sealed class SubstationSystem : EntitySystem 
+{
+    
+    [Dependency] private readonly PointLightSystem _lightSystem = default!;
+    [Dependency] private readonly SharedPointLightSystem _sharedLightSystem = default!;
+    [Dependency] private readonly AppearanceSystem _appearance = default!;
+
+
+    private bool _substationDecayEnabled = true;
```
Comment (0x6273) [inline Content.Server/Power/EntitySystems/SubstationSystem.cs:25]:
```suggestion
{
    [Dependency] private readonly PointLightSystem _lightSystem = default!;
    [Dependency] private readonly SharedPointLightSystem _sharedLightSystem = default!;
    [Dependency] private readonly AppearanceSystem _appearance = default!;

    private bool _substationDecayEnabled = true;
```

File: Content.Server/Power/EntitySystems/SubstationSystem.cs [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+            else
+            {
+                args.PushMarkup(
+                    Loc.GetString("substation-component-examine-malfunction")
+                );
+            }
+            }
+            
+        }
+    }
```
Comment (0x6273) [inline Content.Server/Power/EntitySystems/SubstationSystem.cs:107]: Indentation looks sus here.

**Review by 0x6273 — CHANGES_REQUESTED** [review]

File: Content.Server/Power/EntitySystems/SubstationSystem.cs (hunk 4 above)
Comment (0x6273) [reply Content.Server/Power/EntitySystems/SubstationSystem.cs:25]: Also you should either prefix all or none of these dependencies with `System`, so it's consistent across the file.

File: Content.Client/Power/Substation/SubstationVisualsSystem.cs (hunk 1 above)
Comment (0x6273) [reply Content.Client/Power/Substation/SubstationVisualsSystem.cs:15]: There are tabs here also on the second if. Please make sure your code only uses 4 space indentations.

**Comment by joshepvodka** [comment]
closing to change it into a normal pr

**Pull request closed without merging**